drivers/ina3221/ina3221_saul.c

```diff
@@ -0,0 +1,150 @@
+        return -ECANCELED;
+    }
+    ina3221_channel_t ch = (((ench & INA3221_ENABLE_CH1) ? INA3221_CH1 : 0) |
+                            ((ench & INA3221_ENABLE_CH2) ? INA3221_CH2 : 0) |
+                            ((ench & INA3221_ENABLE_CH3) ? INA3221_CH3 : 0));
+    ina3221_read_shunt_uv(_dev, ch, shunt_uv, NULL);
+    ina3221_calculate_current_ua(_dev, ch, shunt_uv, current);
+    res->scale = -6; /* uA to A*/
+    res->unit = UNIT_A;
+    for (int i = 0, j = 0; i < INA3221_NUM_CH; i++) {
+        res->val[i] = (ch & (1 << i))
+                      ? (int16_t)current[j++] : SAUL_INA3221_NO_VALUE;
+    }
+    phydat_fit(res, current, 3);
+    return INA3221_NUM_CH;
+}
+
+static int read_power(const void *dev, phydat_t *res)
+{
+    ina3221_t *_dev = (ina3221_t *)dev;
+    ina3221_enable_ch_t ench = 0;
+    int32_t shunt_uv[INA3221_NUM_CH] = { 0 };
+    int32_t current_ua[INA3221_NUM_CH] = { 0 };
+    int16_t bus_mv[INA3221_NUM_CH] = { 0 };
+    int32_t power[INA3221_NUM_CH] = { 0 };
+    int num_ch = _ina3221_get_enable_channel(_dev, &ench);
+
+    if (!num_ch) {
+        return -ECANCELED;
+    }
+    ina3221_channel_t ch = (((ench & INA3221_ENABLE_CH1) ? INA3221_CH1 : 0) |
+                            ((ench & INA3221_ENABLE_CH2) ? INA3221_CH2 : 0) |
+                            ((ench & INA3221_ENABLE_CH3) ? INA3221_CH3 : 0));
+    int num_ch_shunt_uv = ina3221_read_shunt_uv(_dev, ch, shunt_uv, NULL);
+    ina3221_calculate_current_ua(_dev, ch, shunt_uv, current_ua);
+    ina3221_read_bus_mv(_dev, ch, bus_mv, NULL);
+    ina3221_calculate_power_uw(bus_mv, current_ua, num_ch_shunt_uv, power);
+    res->scale = -6; /* uW to W*/
+    res->unit = UNIT_W;
+    for (int i = 0, j = 0; i < INA3221_NUM_CH; i++) {
+        res->val[i] = (ch & (1 << i))
+                      ? (int16_t)power[j++] : SAUL_INA3221_NO_VALUE;
+    }
+    phydat_fit(res, power, 3);
+    return INA3221_NUM_CH;
+}
+
+int read_shunt_voltage_sum(const void *dev, phydat_t *res)
+{
+    ina3221_t *_dev = (ina3221_t *)dev;
+    ina3221_enable_ch_t ench = 0;
+    int32_t shunt_voltage_sum = SAUL_INA3221_NO_VALUE;
+    int num_ch = _ina3221_get_enable_channel(_dev, &ench);
+
+    if (!num_ch) {
+        return -ECANCELED;
+    }
+    ina3221_read_shunt_sum_uv(_dev, &shunt_voltage_sum, NULL);
+    res->scale = -6; /* uV to V*/
+    res->unit = UNIT_V;
+    res->val[0] = shunt_voltage_sum;
+    phydat_fit(res, &shunt_voltage_sum, 1);
+    return 1;
+}
+
+const saul_driver_t ina3221_saul_drivers[] = {
+    (saul_driver_t) {
+        .read = read_bus_voltage,
+        .write = saul_notsup,
+        .type = SAUL_SENSE_VOLTAGE
+    },
+    (saul_driver_t) {
+        .read = read_current,
+        .write = saul_notsup,
+        .type = SAUL_SENSE_CURRENT
+    },
+    (saul_driver_t) {
+        .read = read_power,
+        .write = saul_notsup,
+        .type = SAUL_SENSE_POWER
+    },
+    (saul_driver_t) {
+        .read = read_shunt_voltage_sum,
+        .write = saul_notsup,
```

Comment:
I'm not sure if this is a good idea, but I just drop this here to let you decide on this:

How about adding a write function to control the enabled channels. E.g. you could say that 1 (`== 0b001`) is channel 1, 2 (`== 0b010`) is channel two, and 4 (`==0b100`) is channel three. If I would write 0, all channels would be disabled for the sum. Writing 5 (`== 0b101`) would enable 1 and 3, but not channel 2. etc.

Reply:
I think this is a great idea. I will add this soon.